sync w/ tags from wiki, add tags to all missing scripts

p3r7 opened the pull request:
for the norns.community (documentation) 2.0 rework.

- i extracted a parsed index from wiki.js: [wiki_js_norns_community.json](https://gist.github.com/p3r7/2f61ec64954935ab60eae47e828d14ac)
- i did a small script to diff the 2 and also to replace/merge the tags [community_json_add_tags.py](https://gist.github.com/p3r7/85dcd732d5cb822ef93d5cfde2890a6e)
- i went through all the scripts not in the wiki and added appropriate tags for them

there are still quite a few scripts that were on the wiki but not in this JSON file.

ngwese commented:
This change mixes singular and plural words as tag name. The current convention is singular words for tag names which I would advocate retaining.

p3r7 commented:
agreed to make any change (as of now those are the tags as they got defined on the wiki.js but it's a good occasion to review those).

no sure if you both are referring to the same thing for singular/plural.

is that singular as in *"no S at the end"* or singular as in *"no space as separator"*?

ngwese commented:
I only glanced briefly, my comment was about things ending in “s” like “sequencers”. If there are other tags which a compound phrases or multiple words I guess I would advocate breaching those up. I don’t know if the catalogs have tags like that (hopefully not)

ngwese commented:
Also looking more closely there are a few cases where tags that exist i. the catalog are getting dropped in favor of tags from the wiki - would suggest merging as opposed to treating the wiki as wholly authoritative.

p3r7 commented:
(NB: sorry if my wording may seem pompous, i'm not a native speaker)


## the `delays + loopers` edge-case

looking more closely at it, in looks like in the wiki script are still in fact tagged either `delay`, `looper` or both but a decision was made to merge the 2 after the fact as to not demultiply the categories.

i could easily split back this particular tag value but we'd have to bring the intelligence of merging them in a the new static catalog website thingy in the works.

summoning @tyleretters who was there when categories got decided for the wiki and who is also toying w/ the new implem.


## few cases of tags that exist in the catalog are getting dropped in favor of tags from the wiki

this was a voluntary (but not definitive) move.

as soon as those tags are not simple "hashtags" but meta-data used for categorization we'd have to limit the # of possible values demultiply, which can make information management a hassle. (Tyler colloquially referred to this as "tag soup").  

in fact, i originally wanted to **not use** this `tags` field to store the meta-data for script categories & IO capabilities, but create 2 new fields.

Brian agreed w/ the approach but wasn't 100% confident that maiden would be resilient to adding new fields... so ultimately we thought of reusing the existing `tags` one.

@ngwese maybe you could offer a better vision of feasibility of adding fields?

tyleretters commented:
i'd start from the top down with the overall heuristics and conventions. draft:

- tags MUST be a single word
- tags MUST contain no spaces
- tags MUST be singular (no "s" at the end)
- tags MUST contain only alpha-numeric characters
- tags MUST be lower case
- tags MUST NOT contain special characters
- tags SHOULD NOT be acronyms, except in special cases like MIDI

---

some of the confusion probably arose from the wiki.js implementation. tags are automatically pulled into the navigation. from a UX standpoint, the nav ought to say "sequencers" instead of "sequencer" because, well, you're looking at all the sequencers, not just one.

---

and delays and loopers are different things imo

p3r7 commented:
thanks Tyler, for chiming in & notably the clarification regarding looper / delay.

unfortunately, on that front, regarding my comment above:

>  in the wiki script are still in fact tagged either delay, looper or both

this is in fact wrong. i got misguided by this piece of code that does the tags -> categories for the gallery view:

![image](https://user-images.githubusercontent.com/11557146/221288201-209aaac0-8076-4829-821a-e6428fabd6fa.png)
 
but in fact this does nothing right now as it appears the 2 tag values got [merged in a single tag in wiki.js itself](https://norns.community/t/delays%20+%20loopers?sort=title).

![image](https://user-images.githubusercontent.com/11557146/221289022-4c780947-3291-4b37-91f7-45756c66d6b9.png)

![image](https://user-images.githubusercontent.com/11557146/221288858-2924cd2b-ef0f-4d98-b28b-1ed0ec5dea6c.png)

for now i've changed the tags to be singular words and have replaced spaces w/ "_". i'll have to go through each one tagued "delay_looper" and retag them manually from their previous tags + info in lines threads.

p3r7 commented:
ok, split the delay & looper tags.

(also removed the tag from 1 or 2 scripts where it was mistagged)

but while doing this, i start to understand why the 2 tags got merged: any looper that allows live overdubbing can be used as a delay. so a lot of the scripts, esp. those built w/ softcut, fit into both categories at once.

p3r7 commented:
also just did a merge of most of the tags from current version.

a few script had only some of their IO capabilities defined on the wiki and the others here which was weird. but merging the 2 should give a more complete view.

also enabled back the `noise`, `drone` & `tracker` tags on some scripts. those sound oddly specific but clear boundaries on categories, even broad ones, are hard if not impossible to define anyway.

in fact the deciding factor to make a category "appear" in the gallery is when it reached a critical mass of enough scripts falling into it.

ngwese reviewed (approved):
LGTM

ngwese commented:
the original idea behind the tags was to facilitate implementing some search features in `maiden` but that was never implemented. i think there is going to always be tension between labels that seem applicable/descriptive from the script author and labels that are viewed as desirable for curation purposes (like navigation on a website).

the tag criteria which @tyleretters outlined above largely match the undocumented spirit behind the tags field in the first place.

fields can be added IIRC without breaking `maiden` if memory serves but it has been long enough that i don't know with absolute certainly. @p3r7 was the idea to take what is in `tags` today split that into a `categories` and `io` fields? ...so "sequencer" would be considered a category but "midi" would be considered io?

p3r7 commented:
> was the idea to take what is in `tags` today split that into a `categories` and `io` fields? ...so "sequencer" would be considered a category but "midi" would be considered io?

yes. precisely. but to also keep the tags field for an eventual less "typed" & rigorously-enforced meta-data.

how does  that sounds to you?

(i'm fine w/ keeping everything under `tags` as well)


> fields can be added IIRC without breaking maiden if memory serves

i didn't realize that maiden was written in Go, as highlighted by [your comment in the previous PR](https://github.com/monome/norns-community/pull/246#issuecomment-1444996425).

indeed, iirc `json.Unmarshal` won't complain on unknown filed unless using a custom decoder initialized w/ the `.DisallowUnknownFields()` flag, at least on recent go version.

(and of course the yaml parser w/ a similar API has the opposite default behavior...)

doing a manual test to confirm all this won't hurt.

ngwese commented:
taking a step back for a second my understanding is that the goal here is to create a static site generator to facilitate browsing through scripts off device.

moving data from tags to new fields will result in the maiden catalog view loosing information that is there today which I’m not in favor of. the maiden web app could be updated to respect/merge stuff from the new fields but the associated catalog change would break compatibility with older devices not on the latest software which is also not good. the values could be duplicated across multiple fields but that in my opinion invites confusion and additional maintenance.

the desire to split what is tags today into “category” and “io” strikes me as curation which is focused on the site generation as it is being imagined today. what if a different navigation structure is desired tomorrow? what if authors add lots of specific tags/categories which creates noise in the generated site navigation? at what point does a maybe overly specific tag (“tracker”) become a category for site navigation?

the way i would approach this is to leave that catalog as is and write a script which consumes a catalog file and enriches it with the desired information which the site generator wants. for example:
- test for tags which match identified “well known” categories (i.e. sequencer or sampler) or io and generate the additional fields accordingly
- look for tags which are not in the well known list but occur on more than 5 scripts (for example) and treat that as an indication of an emergent category

ngwese commented:
going ahead and merging this so that future contributions don't result in conflicts

tyleretters commented:
thanks for doing this @p3r7!! huge amount of work. i very much appreciate it. <3

p3r7 commented:
thanks @ngwese for the merge but more importantly all the feedback.

> occur on more than 5 scripts (for example) and treat that as an indication of an emergent category

i *really* like this idea. auto-scaling of the categorization.

tyleretters commented:
is this a substantial enough change for a bump to v2?

https://github.com/p3r7/norns-community/blob/3d83207627b135fc8f3eb58ba9a1d55e703b12e5/community.json#L3

ngwese commented:
the version number is for the file format which isn’t changing with this PR so no need to change it
